lms/djangoapps/courseware/rules.py

```diff
@@ -24,3 +29,80 @@ def is_track_ok_for_exam(user, exam):
 # proctored experience
 can_take_proctored_exam = is_track_ok_for_exam
 rules.set_perm('edx_proctoring.can_take_proctored_exam', is_track_ok_for_exam)
+
+
+class HasAccessRule(Rule):
+    """
+    A rule that calls `has_access` to determine whether it passes
+    """
+    def __init__(self, action):
+        self.action = action
+
+    def check(self, user, instance=None):
+        return has_access(user, self.action, instance)
+
+    def query(self, user):
+        raise NotImplementedError()
+
+
+class IsStaffInCourse(Rule):
+    """
+    Check whether a user has `staff` access in a course.
+
+    Expects to be used to filter a CourseOverview queryset
+    """
+    def check(self, user, instance=None):
+        staff_sql_experiment = laboratory.Experiment(raise_on_mismatch=settings.DEBUG)
+        staff_sql_experiment.control(self._check_with_has_access, args=(user, instance))
+        staff_sql_experiment.candidate(self._check_with_query, args=(user, instance))
+        return staff_sql_experiment.conduct()
+
+    def _check_with_has_access(self, user, instance=None):
+        return has_access(user, 'staff', instance)
+
+    def _check_with_query(self, user, instance=None):
+        # delegate the work to type-specific functions.
+        # (start with more specific types, then get more general)
+        if isinstance(instance, (CourseDescriptor, CourseOverview)):
+            course_key = instance.id
+        elif isinstance(instance, (ErrorDescriptor, XModule, XBlock)):
+            course_key = instance.scope_ids.usage_id.course_key
+        elif isinstance(instance, CourseKey):
+            course_key = instance
+        elif isinstance(instance, UsageKey):
+            course_key = instance.course_key
+        elif isinstance(instance, basestring):
+            course_key = CourseKey.from_string(instance)
+
+        return self.filter(user, CourseOverview.objects.filter(id=course_key)).exists()
+
+    def query(self, user):
+        if not user.is_authenticated:
+            return EMPTY
+
+        masq_settings = getattr(user, 'masquerade_settings', {})
+        masq_as_student = [
+            course_key for
+            (course_key, masq_setting) in masq_settings.items()
+            if masq_setting.role == 'student'
+        ]
+
+        not_masquerading_as_student = ~Q(id__in=masq_as_student)
+
+        is_global_staff = user.is_staff
+        course_staff_or_instructor_courses = CourseAccessRole.objects.filter(
+            user=user,
+            role__in=('staff', 'instructor'),
+            course_id__isnull=False
+        ).values('course_id')
+        org_staff_or_instructor_courses = CourseAccessRole.objects.filter(
+            user=user,
+            role__in=('staff', 'instructor'),
+            course_key__isnull=True,
+            org__isnull=False
+        ).values('org')
+
+        query = not_masquerading_as_student
```

Comment:
why is this check here?  
it makes sense to me but i'm having trouble thinking through the masquerade business logic edge cases in my head and why we'd check for student masquerade but not other masquerade for example

Reply:
It's that way because that's how `has_access` is written.